### ISSUE TITLE
Support CORS and OPTION requests in API

### ISSUE THREAD

**Issue by pdesjardins90** [forem/forem#4195]
**Is your feature request related to a problem? Please describe.**

I'm trying to dynamically display previews of my latest blog posts on my personal website with links to the originals on DEV, but querying the API with fetch results in a CORS error. (Works fine in postman)

Error:

*"Access to fetch at 'https://dev.to/api/articles?username=pdesjardins90' from origin 'https://philippe-desjardins.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status."*

**Describe the solution you'd like**

I'd like to be able to make fetch requests to the DEV API from another host (with OPTIONS pre-flight requests) without any issue.

**Describe alternatives you've considered**

Not querying the DEV API in the browser and just hardcoding my blog posts previews and update them manually every time I write one. :(

**Comment by zevaverbach** [comment]
👋 I'm taking a crack at this today.

**Comment by zevaverbach** [comment]
Okay to implement this for videos and podcasts while I'm at it?

**Comment by rhymes** [comment]
@zevaverbach I think this issue is for the entirety of the API, but we might want to discuss over here or over there (once your PR is submit) what kind of access to give each resource. 

One thing: CORS should only be enabled on public API endpoints, those that don't require any type of authentication

**Comment by zevaverbach** [comment]
@rhymes @jessleenyc Let me know if this is the wrong place to ask, but I'm hitting a wall after many hours of trying to get this running locally. `bin/startup` and `docker-compose up` both fail after following all the guidelines, setting up the db, etc., as well as applying a required migration.

I'm happy to keep debugging this on my own, but can you point me in the right direction for getting useful logs?

More details:

With docker, there's a lot of similar errors related to [`relation 'delayed jobs' does not exist`](https://gist.github.com/zevaverbach/d4b7f9545d1a877acc3dd8732f2f2d64)

In plain MacOS (`bin/startup`) this is happening when trying to access at `localhost:3000`:

```
15:23:00 web.1       | objc[21478]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called.
15:23:00 web.1       | objc[21478]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
15:23:01 web.1       | objc[21477]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called.
15:23:01 web.1       | objc[21477]: +[__NSPlaceholderDate initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

**Comment by zevaverbach** [comment]
Hm, this may have been premature, trying `export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`...

**Comment by rhymes** [comment]
@zevaverbach what version of macOS do you have?

What happens if you open three tabs, and start separately `./bin/webpack-dev-server`, `rails s` and `rails jobs:work` ?

**Comment by rhymes** [comment]
Also, what version of Ruby are you using?

**Comment by zevaverbach** [comment]
@rhymes MacOS 10.14.5, Ruby 2.6.3p62. Will try your suggestion next if this `export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` doesn't work.

**Comment by zevaverbach** [comment]
@rhymes okay, I'm in business with `export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES` -- should I make a separate PR to add that to the docs somewhere?!

**Comment by rhymes** [comment]
I have the same exact setup but I've never noticed that, you could add it to the "possible error messages" section of https://github.com/thepracticaldev/dev.to/blob/master/docs/installation/mac.md#possible-error-messages in a separate PR

**Comment by zevaverbach** [comment]
Will have something today.

**Comment by zevaverbach** [comment]
@pdesjardins90 I'm having trouble replicating this issue. Can you provide the exact fetch call you're making, please?

**Comment by zevaverbach** [comment]
@pdesjardins90 I made a tiny test [here](https://github.com/zevaverbach/zevaverbach.github.io/blob/08eab9c9db2c66b43ab8968f3389092745325884/index.html#L134-L138) using my own Dev.to username, and as you can (hopefully) see in the console [here](https://averba.ch), there's no cors error.

**Comment by zevaverbach** [comment]
@pdesjardins90 just pinging you to see if you have any thoughts.

**Comment by pdesjardins90** [comment]
@zevaverbach just came back from a two-week vacation, sorry for the delay! Just tried it on my end, same setup as I had and now it works, someone fixed in somehow since I guess? Anyway thanks for the help!

**Comment by rhymes** [comment]
Reopening because the issue is applicable to all endpoints, they have not been verified to be CORS enabled.

We should keep this open as a bookmark and to invite further discussion on how to solve this once and for all

**Comment by rhymes** [comment]
@pdesjardins90 Sorry, it took us a while to get back to fix this. Can you confirm it works for you?

I tried with:

```js
response = await fetch("https://dev.to/api/articles?username=pdesjardins90", { headers: { "Origin": "https://philippe-desjardins.com/" }})
await response.json()
```

and it worked well :)

**Comment by pdesjardins90** [comment]
@rhymes yeah it does! I tried closing this end of october as you can see up there but someone said it had to be verified for all endpoints. Feel free to close it now if everything's fine, thanks for the effort!

**Comment by rhymes** [comment]
Thanks for confirming, I'm going to leave this open as we haven't enabled it for authenticated endpoints yet :)

**Comment by cmgorton** [comment]
@rhymes do you know where we are at with this issue? Is this something we should label as a potential RFC?

**Comment by rhymes** [comment]
@cmgorton thanks for bringing this up! I don't think we can dedicate bandwidth to this now as it sits in a much larger conversation about how to evolve the API. In addition to that and on very practical terms, we don't want to enable CORS for authenticated endpoints as we require an `api-key` header which would likely end up being embedded in public frontends with understandable consequences. 

I think we'll revisit all of this once we can properly focus on evolving Forem's API

**Comment by cmgorton** [comment]
Ok I am going to close this issue for now and we can reopen it when we have time to dedicate an RFC for it.

**Comment by vatana7** [comment]
```javascript
await fetch("https://dev.to/api/articles/me", {
    headers: {
      "api-key": token
    },
    method:"GET",
    mode:"cors"
  }).then((res) => console.log(res));
```

I still have this problem. I'm using `axios` in React App using Vite
